doc/src/devdocs/llvm.md

```diff
@@ -75,3 +75,159 @@ study it and the pass of interest in isolation.
+This is problematic, because it breaks that stated requirement that we're able
+to identify gc-relevant pointers. This invariant is accomplished using the LLVM
+"non-integral pointers" feature, which is new in LLVM 5.0. It prohibits the
+optimizer from making optimizations that would introduce these operations. Note
+we can still insert static constants at JIT time by using inttoptr in address
+space 0 and then decaying to the appropriate address space afterwards.
+
+### Supporting ccall
+One important aspect missing from the discussion so far is the handling of
+`ccall`. `ccall` has the peculiar feature that the location and scope of a use
+do not coincide. As an example consider:
+```
+A = randn(1024)
+ccall(:foo, Void, (Ptr{Float64},), A)
+```
+In lowering, the compiler will insert a conversion from the array to the
+pointer which drops the reference to the array value. However, we of course
+need to make sure that the array does stay alive while we're doing the ccall.
+To understand how this is done, first recall the lowering of the above code:
+```
+    return $(Expr(:foreigncall, :(:foo), Void, svec(Ptr{Float64}), :($(Expr(:foreigncall, :(:jl_array_ptr), Ptr{Float64}, svec(Any), :(A), 0))), :(A)))
+```
+The last `:(A)`, is an extra argument list inserted during lowering that informs
+the code generator which julia level values need to be kept alive for the
+duration of this ccall. We then take this information and represent it in an
+"operand bundle" at the IR level. An operand bundle is essentially a fake use
+that is attached to the call site. At the IR level, this looks like so:
+```
+    call void inttoptr (i64 ... to void (double*)*)(double* %5) [ "jl_roots"(%jl_value_t addrspace(10)* %A) ]
+```
+The GC root placement pass will treat the jl_roots operand bundle as if it were
+a regular operand. However, as a final step, after the gc roots are inserted,
+it will drop the operand bundle to avoid confusing instruction selection.
+
+### Supporting pointer_from_objref
+`pointer_from_objref` is special because it requires the user to take explicit
+control of GC rooting. By our above invariants, this function is illegal,
+because it performs an addressspace cast from 10 to 0. However, it can be useful,
+in certain situations, so we provide a special intrinsic:
+```
+declared %jl_value_t *julia.pointer_from_objref(%jl_value_t addrspace(10)*)
+```
+which is lowered to the corresponding address space cast after gc root lowering.
+Do note however that by using this intrinsic, the caller assumes all responsibility
+for making sure that the value in question is rooted. Further this intrinsic is
+not considered a use, so the GC root placement pass will not provide a GC root
+for the function. As a result, the external rooting must be arranged while the
+value is still tracked by the system. I.e. it is not valid to attempt use the
+result of this operation to establish a global root - the optimizer may have
+already dropped the value.
```

Comment:
This escape semantic would also be required so that we can stack allocate `RefValue{bitstype}` in ccall.